src/python/pants/engine/internals/target_adaptor.py

```diff
@@ -0,0 +1,40 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Any
+
+from typing_extensions import final
+
+
+@final
+class TargetAdaptor:
+    """A light-weight object to store target information before being converted into the Target
+    API."""
+
+    def __init__(self, type_alias: str, name: str, **kwargs: Any) -> None:
+        self.type_alias = type_alias
+        self.name = name
+        self.kwargs = kwargs
+
+    def _key(self):
+        def hashable(value):
+            if isinstance(value, dict):
+                return tuple((k, hashable(v)) for k, v in value.items())
+            if isinstance(value, list):
+                return tuple(hashable(v) for v in value)
+            if isinstance(value, set):
+                return tuple(sorted(hashable(v) for v in value))
+            return value
+
+        return (self.type_alias, self.name, *sorted(hashable(self.kwargs)))
+
+    def __hash__(self):
+        return hash(self._key())
+
+    def __eq__(self, other):
+        if not isinstance(other, TargetAdaptor):
+            return NotImplemented
+        return self._key() == other._key()
```

Comment:
I'm not actually sure but: is this ever actually used in a way that requires `__hash__`/`__eq__`? If not, you might be able to disable them instead.

Reply:
Turns out, they are not.. Cool, deleting.

Reply:
How did you confirm that `__eq__` is not used? ... sorry, I might have lead you slightly astray.

`__eq__` is used in more places than `__hash__`: we use it to determine whether a `@rule`'s output has changed between runs, and to "clean" (ie, skip re-running) nodes if it hasn't. But `__eq__` also has a default implementation that is not helpful for that case.

Reply:
I added `raise ValueError()` to both dunders, then ran `./pants dependencies ::`. Neither method was used.

Reply:
I added back a simple `__eq__`, which I agree we need and a unit test was breaking without.

This means that `__hash__` defaults to None. Do we need `__hash__` implemented too here? We do have a rule that `-> TargetAdaptor`. Are you saying that we should have only been using the `_key` function in `__hash__`, and `__eq__` could have been much simpler?

Reply:
> Do we need __hash__ implemented too here?

`__hash__` will fail noisily if it is not implemented, so you're always safe to not have it unless you determine that some codepath needs it. Using a value as a `Param` (at the root of the graph, or as the input to a `Get`) requires both `__hash__` and `__eq__`. If something is not being used as a `Param`, then `__eq__` only comes into play to improve caching behavior, as described in https://github.com/pantsbuild/pants/pull/10309#discussion_r452597829